[UNOMI-828] Support for OpenSearch persistence

sergehuber opened the pull request:
Dear Apache Unomi developers and users,

After 8 months of development, and while I'm at it to start 2025 on a high note, I am proud to share with you and submit for review my contribution on the support of OpenSearch as an alternative to ElasticSearch for Apache Unomi.

A lot of work has gone into this contribution, and I can fully understand that some of you will have questions. I propose to organize a meeting to go over the changes and also to answer any questions you might have.

LIVE DEMONSTRATION ANNOUNCEMENT
-----------------------------
I'm pleased to invite you to a live presentation and demonstration of the OpenSearch integration and new development tools. 

Date: January 13th, 2025
Time: 17:00 CET
Format: Online presentation with live demonstration
Link: https://calendar.app.google/7ZffzAgwPwhamAAU6

Agenda:
1. Overview of OpenSearch integration
2. Demonstrations:
   - Configuring and starting Apache Unomi with OpenSearch
   - New development tools in action
   - Integration testing capabilities
3. Q&A session

OPENSEARCH CONTRIBUTION RATIONALE
---------
- Enable true vendor independence by providing a fully open-source search engine alternative
- Address growing community concerns about ElasticSearch licensing changes and potential future restrictions
- Reduce total cost of ownership for Apache Unomi deployments, especially in large-scale environments
- Future-proof Apache Unomi's architecture through a modular persistence layer design
- Enhance developer experience with flexible backend choices
- Leverage OpenSearch's strong commitment to the Apache License 2.0
- Enable seamless scaling without licensing constraints
- Benefit from OpenSearch's active development and innovation in the search engine space
- Strengthen Apache Unomi's position as a truly free and open-source CDP platform
-  Support the latest and greatest version of OpenSearch
- Don't break existing ElasticSearch support for existing users
- Use existing integration tests to validate contribution for both backend implementations

WHERE TO FIND IT
——————————

https://github.com/apache/unomi/tree/opensearch-persistence

WHAT’S INCLUDED
—————————
- 100% feature parity between ElasticSearch and OpenSearch
- 100% integration tests passing on both ElasticSearch and OpenSearch
- Updates to in-project documentation (website updates will be done once these changes are reviewed)

WHAT IS NOT INCLUDED
————————————
- Migration from ElasticSearch to OpenSearch: this contribution is aimed at new projects that are starting to deploy Apache Unomi. That doesn’t mean to say that migration is not possible or even difficult (apart from data size), it just wasn’t in the initial scope of this contribution. 

DETAILED CHANGES SUMMARY
-----------------------

I'd like to summarize the significant improvements and changes implemented in the opensearch-persistence branch. This work represents a major enhancement to Apache Unomi's persistence layer and development tooling.

Key Features and Improvements:

1. OpenSearch Integration
   - Full support for OpenSearch 2.18.0 as an alternative to ElasticSearch
   - Complete implementation of persistence layer for OpenSearch
   - Support for authenticated and encrypted OpenSearch configurations
   - Implementation of roll-over policy for OpenSearch
   - Backend-independent implementation of GeoDistance and DateMath utilities
- Integration with Health Check servlet

2. Testing and CI/CD Improvements
   - Integration tests now fully compatible with both OpenSearch and ElasticSearch
   - Docker-based OpenSearch testing environment
   - Prepared GitHub Workflows matrix (pending testing)
   - New integration tests slowest tests report for performance monitoring
   - Maven Cache integration for faster builds
   - Support for Karaf Debugging in Docker environments

3. Architecture and Configuration
   - Feature-based persistence implementation selection
   - Configurable startup process via org.apache.unomi.start.cfg
   - Clean separation between OpenSearch and ElasticSearch deployments
   - Split Unomi features into smaller, more manageable chunks
   - Enhanced configuration options for both persistence implementations

4. Development Tools and Scripts
   - New build script with comprehensive options for:
     * Building and running the application
     * Running integration tests with either ElasticSearch or OpenSearch
     * Maven Cache management integration (including purging/bypassing)
   - New Karaf Tools helper script for integration tests:
     * Log file viewing and tailing
     * Container management (start/stop)
     * Debug mode support
     * Grep functionality for logs

5. Command Line Interface Improvements
   - Enhanced unomi:start command with persistence selection
   - New shell commands for resource management including removal of items (session, profiles, actions, …)
   - Improved error handling and user feedback
   - Added confirmation prompts for destructive operations

6. Code Quality and Maintenance
   - Improved logging consistency
   - Enhanced error handling
   - Better resource cleanup
   - Updated servlet configurations
   - Version bump to 2.7.0-SNAPSHOT

Test Performance Insights:
The slowest test analysis reveals areas for potential optimization, with some tests taking up to 100 seconds to complete. This information will be valuable for future performance improvements. Also a intermediate progress report is now printed indicating how many tests have succeeded or failed, as well as a time estimation until they are completed and a progress bar.

Next Steps:
1. Code review and merge of the branch
2. Complete testing of the GitHub Workflows matrix
3. Documentation updates to reflect new features and configurations
4. Release new version
5. Update website with latest OpenSearch support information

The changes represent a significant step forward in making Apache Unomi more flexible and maintainable, while providing developers with better tools for development and debugging.

----

**Please** following this checklist to help us incorporate your contribution quickly and easily:

 - [X] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/UNOMI) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [X] Format the pull request title like `[UNOMI-XXX] - Title of the pull request`
 - [X] Provide integration tests for your changes, especially if you are changing the behavior of existing code or adding
       significant new parts of code.
 - [X] Write a pull request description that is detailed enough to understand what the pull request does, how, and why. 
       Copy the description to the related JIRA issue
 - [X] Run `mvn clean install -P integration-tests` to make sure basic checks pass. A more thorough check will be 
        performed on your pull request automatically.
 
Trivial changes like typos do not require a JIRA issue (javadoc, project build changes, small doc changes, comments...). 
 
If this is your first contribution, you have to read the [Contribution Guidelines](https://unomi.apache.org/contribute.html)

If your pull request is about ~20 lines of code you don't need to sign an [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) 
if you are unsure please ask on the developers list.

To make clear that you license your contribution under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [X] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

Fgerthoffert commented:
Hello,

Thanks a lot for the huge work on this PR, as mentioned during the demo, we plan to dedicate some time to review the pull request starting from January 31st (or earlier if we can).

We are doing our best to chime asap to avoid the PR dragging for too long.

sergehuber commented:
Thank you @Fgerthoffert that is much appreciated. Also, if I can help maybe by directly briefing reviewers on the changes in the code that's something I can do.

Best regards,
  Serge.

jayblanc commented:
I started a review

jayblanc commented on persistence-opensearch/core/src/main/java/org/apache/unomi/persistence/opensearch/OpenSearchPersistenceServiceImpl.java:
```diff
@@ -0,0 +1,2761 @@
+        return Collections.singletonMap(property.getItemId(), definition);
+    }
+
+    private String convertValueTypeToESType(String valueTypeId) {
```
Change ES to OS

jayblanc commented on persistence-opensearch/core/src/main/java/org/apache/unomi/persistence/opensearch/OpenSearchPersistenceServiceImpl.java:
```diff
@@ -0,0 +1,2761 @@
+                return "date";
+            case "string":
+            case "id":
+            case "email": // TODO Consider supporting email mapping in ES, right now will be map to text to avoid warning in logs
```
Use of ES instead of OS

sergehuber replied:
```suggestion
            case "email": // TODO Consider supporting email mapping in OS, right now will be map to text to avoid warning in logs
```

sergehuber commented on persistence-opensearch/conditions/src/main/java/org/apache/unomi/persistence/opensearch/conditions/PropertyConditionOSQueryBuilder.java:
```diff
@@ -0,0 +1,255 @@
+
+    private void checkRequiredValuesSize(Collection<?> values, String name, String operator, int expectedSize) {
+        if (values == null || values.size() != expectedSize) {
+            throw new IllegalArgumentException("Impossible to build ES filter, missing " + expectedSize + " values for a condition using comparisonOperator: " + operator + ", and propertyName: " + name);
```
```suggestion
            throw new IllegalArgumentException("Impossible to build OS filter, missing " + expectedSize + " values for a condition using comparisonOperator: " + operator + ", and propertyName: " + name);
```

sergehuber commented on persistence-opensearch/conditions/src/main/java/org/apache/unomi/persistence/opensearch/conditions/PropertyConditionOSQueryBuilder.java:
```diff
@@ -0,0 +1,255 @@
+        String name = (String) condition.getParameter("propertyName");
+
+        if (comparisonOperator == null || name == null) {
+            throw new IllegalArgumentException("Impossible to build ES filter, condition is not valid, comparisonOperator and propertyName properties should be provided");
```
```suggestion
            throw new IllegalArgumentException("Impossible to build OS filter, condition is not valid, comparisonOperator and propertyName properties should be provided");
```

sergehuber commented on persistence-opensearch/conditions/src/main/java/org/apache/unomi/persistence/opensearch/conditions/PropertyConditionOSQueryBuilder.java:
```diff
@@ -0,0 +1,255 @@
+
+    private void checkRequiredValue(Object value, String name, String operator, boolean multiple) {
+        if (value == null) {
+            throw new IllegalArgumentException("Impossible to build ES filter, missing value" + (multiple ? "s" : "") + " for condition using comparisonOperator: " + operator + ", and propertyName: " + name);
```
```suggestion
            throw new IllegalArgumentException("Impossible to build OS filter, missing value" + (multiple ? "s" : "") + " for condition using comparisonOperator: " + operator + ", and propertyName: " + name);
```

jkevan reviewed (changes requested):
Overall the works related to open search support looks good and promising.

I am a bit surprised by the quantity of addition/modifications that are not related to open search support directly. Due to the complexicity of the changes, and the tons of files it involved I would be in favor separating your PR in multiple ones, with dedicated tickets for fixes and quality of life improvements. And this for multiple reasons:

- Would be easier to review
- Would be easier to implem test for those new modified parts, I see the code contains some bug fixes, but I dont see any new integrations tests related to them.
- Would be easier to track in the history and link new features/commands/apis in dedicated ticket (since we are using JIRA ticket to generate change logs.)

One down side I see, is the refactoring of the packaging from kar to better decoupling in features, will impacts projects that are building/repackaging Unomi that rely on existing features/kar naming.

Appart from that I added comments in few places, mostly regarding missing java docs or small improvements to consider.

Thanks a lot for the works ! 👍

sergehuber commented on persistence-spi/src/main/java/org/apache/unomi/persistence/spi/conditions/datemath/DateMathParser.java:
```diff
@@ -0,0 +1,270 @@
+import java.util.function.Function;
+import java.util.function.LongSupplier;
+
+public class DateMathParser {
```
I couldn't find a 100% compatible DateMathParser that would work exactly the same way as the Elasticsearch one, which is why this was done. But I agree the Javadoc is missing I will add it.

sergehuber commented on persistence-spi/src/main/java/org/apache/unomi/persistence/spi/CustomObjectMapper.java:
```diff
@@ -58,6 +58,7 @@ public CustomObjectMapper() {
     public CustomObjectMapper(Map<Class, StdDeserializer<?>> deserializers) {
         super();
         super.registerModule(new JaxbAnnotationModule());
+        super.registerModule(new com.fasterxml.jackson.datatype.jsr310.JavaTimeModule());
```
Actually this was required when working with more recent JDKs.

sergehuber commented on persistence-spi/src/main/java/org/apache/unomi/persistence/spi/conditions/ConditionEvaluatorDispatcher.java:
```diff
@@ -0,0 +1,32 @@
+
+import java.util.Map;
+
+public interface ConditionEvaluatorDispatcher {
```
Indeed, will add it.

sergehuber commented on persistence-spi/src/main/java/org/apache/unomi/persistence/spi/conditions/datemath/JavaDateFormatter.java:
```diff
@@ -0,0 +1,351 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class JavaDateFormatter {
```
I agree the context is missing. Basically the idea was to have a 100% compatible implementation with the existing Elasticsearch ones. Introducing new libraries here would have introduced new issues.

sergehuber commented:
Hello @jkevan thank you for your review. 

I will look at your feedback but to answer some of your questions globally : 
- I am aware there are some things that don't seem directly related to the OpenSearch work, however ALL of them were necessary to get OpenSearch to work properly, and some where tools that were developped while working on the project to make sure everything worked. It made more sense to include them in the project since separating them would take more time. I do understand the fact that your preference is to have individual changes but on large modifications such as this this actually wastes time that would be better used at moving the project forward.

- I will add some JavaDocs as I agree some are missing

- About the re-implementation of some classes, these were necessary for the following reasons : stay 100% compatible with the existing implementation. Using other implementations such as the JSR 385 backport would have added possible changes in behavior. A lot of code had to be centralized and therefore new interface had to be created to make sure that the code was properly reusable between the two.

- For the missing tests, actually I have been working on a new REST API test framework that includes testing of some of the new methods but I haven't been sharing that yet. The current integration tests are a nightmare to work with, especially on such a large change.

- About the changes to features, this was not a change I really wanted to make at this point but was forced to do because there was really no clean way to fix the startup with conditional bundles depending on the persistence implementation. The unomi:start implementation was a mess from the start and was already duplicating a lot of the functionality of Karaf features and ideally should be completely removed to just use features. I actually tried to do that but I ran into BIG problems with the pax-web implementation that wouldn't restart properly. This is why I had to set all the bundles to start=false in the feature and then start them otherwise the system wouldn't start properly. I'm hoping this will be fixed when upgrading to a more recent Karaf but this was the only way I managed to get things to work properly for both Elasticsearch and OpenSearch.

- As for the parameter on the unomi:start command this was actually very useful when developing as it made it very easy to start and stop Unomi without starting Karaf to switch between the ElasticSearch and OpenSearch implementations. As it is integrated with the autostart configuration it is also possible to set this up before the start and this has been implemented for the Docker images.

- Did you watch the recording of the meeting I did to present the changes? A lot of the questions you had were presented in that meeting, which I specifically organized to explain the changes and give people opportunities to ask questions and give early feedback. You can find the video here : https://www.youtube.com/watch?v=uNbKW29FCnE . I will be organizing more of these as I am actively working on some more important contributions to the project and I have been sharing with the community at the monthly meetings as well. 

- Last but not least, as I have been working on other major things such as multi-tenancy and other cleanups such as the removal of the clustering code, I am beginning to think that maybe we should move these changes to be part of Unomi V3, which is a branch I have already started working on, and I have also shared my current work in the mailing lists and in multiple meetings. These big changes will slow down quickly however, as I am working on making Unomi much cheaper to deploy and operate. OpenSearch and multi-tenancy are the biggest requirements on this area, along with some cleanups that I would like to do. I am going to communicate some proposals on the subject, but I really need to get these things implemented for my own needs as well as some other users of Unomi. If you are interested in the multi-tenancy, please check my email in the mailing list about it : https://lists.apache.org/list.html?dev@unomi.apache.org

sergehuber commented:
**Changes since February on OpenSearch contribution**
After a LOT of work, I have updated this PR. Hopefully, it will address all the feedback from the reviews and can be merged soon to avoid having to track changes on the master branch again.
Here is a summary of the changes:
- Merged with the latest changes from the master branch, including ES 9 support and Karaf 4.4.8 (tricky merge!)
- Upgraded OpenSearch support to version 3 (was on 2.18 previously)
- Removed all the non-OpenSearch related changes
- Added missing Javadocs
- Addressed all the points raised by the reviewers that were relevant
- Renamed all the _queryBuilderIds_ and removed the *ESQueryBuilder postfix in the queryBuilderIds, replacing them with *QueryBuilder. For example, we had queryBuilderId=booleanConditionESQueryBuilder for both OpenSearch and ElasticSearch, so I simply renamed this to booleanConditionQueryBuilder. In order to avoid introducing a breaking change, I added a system to map old queryBuilderIds to new ones that issues a warning if a legacy ID is being used but still allows the old IDs to work transparently. Also documented the system in the migration guide.
- Developed integration tests for the queryBuilderId mapping system
- Updated the Healthcheck provider to now only report the persistence implementation that is being used
- Documented the new start configuration system and explained how it can be used for custom deployments or distributions
- Removed the Hover event query builder implementation and simply replaced it with a condition type with a parent condition. This way it can work on both ElasticSearch and OpenSearch. Also added a hover event JSON schema built into the plugin.

All the tests are green on GitHub, and the code is ready for review (again)!

sergehuber commented:
Following the release I need to update the version numbers, I'll do this asap.

jayblanc commented on extensions/healthcheck/src/main/java/org/apache/unomi/healthcheck/provider/PersistenceEngineHealthProvider.java:
```diff
@@ -0,0 +1,37 @@
+/*
```
I don't understand the introduction of that interface and I think it is creating too much code for simply muting a healthcheck provider in some configuration conditions.
There is an already existing mechanism to deactivate a health check provider by removing its name from the OSGI config parameter : healthcheck.providers
It is up to the assembly to setup this config.
If it is only a problem of HealthCheckProvider activation, we could let the HealthCheckService manage providers activation (instead of using OSGI lifecycle one) for those that are declared to be active in the config. Thus the OSGI Component could exists but in a unactive state and not being integrated to global healtcheck. Even the provider itself could decide to activate himself by checking in the config if it is part of the declared providers.
To me, this layer of complexity needs to be removed to keep a 1-N relation between the Service and the providers.
If we need to add a behavior in the provider interface to introduce a new orchestration parameter from the service why not but such delegator is not necessary and creates boilerplate code that will be necessary to maintain.

jayblanc commented on extensions/healthcheck/src/main/java/org/apache/unomi/healthcheck/servlet/HealthCheckServlet.java:
```diff
@@ -75,5 +75,6 @@ protected void service(HttpServletRequest request, HttpServletResponse response)
             } else {
                 response.setStatus(HttpServletResponse.SC_PARTIAL_CONTENT);
             }
+        response.flushBuffer();
```
I don't think this is necessary. 
Calling that will send response to client immediately avoiding header to be changed later (in a filter or something else). It is up to the Servlet container to decide when to flush the response buffer.

jsinovassin commented on persistence-spi/src/main/java/org/apache/unomi/persistence/spi/conditions/dispatcher/ConditionQueryBuilderDispatcherSupport.java:
```diff
@@ -0,0 +1,93 @@
+ * {@link #findQueryBuilderKey(String, String, java.util.function.Predicate, org.slf4j.Logger)} methods.
+ * This helper intentionally avoids any dependency on backend-specific query types.
+ */
+public class ConditionQueryBuilderDispatcherSupport {
```
Instead of adding the ConditionQueryBuilderDispatcherSupport, I would suggest to create an abstract class ConditionQueryBuilderDispatcher which will contains findQueryBuilderKey
and an abstract method getLogger()
The ConditionESQueryBuilderDispatcher and ConditionOSQueryBuilderDispatcher could extends ConditionQueryBuilderDispatcher.
